### PR TITLE
Add cases of qcow2 image with native luks encryption

### PR DESCRIPTION
**Pull request by jgao12345**
Add tests of qcow2 image with native luks encryption
```

# avocado run --vt-type libvirt type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.dir_pool_test.target_format_qcow2
JOB ID     : 8891b73993fd0d83260f0b9866c43ed1100c5734
JOB LOG    : /root/avocado/job-results/job-2020-04-02T22.38-8891b73/job.log
 (1/2) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.dir_pool_test.target_format_qcow2.encryption_in_source: PASS (101.63 s)
 (2/2) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.dir_pool_test.target_format_qcow2.encryption_out_source: PASS (184.83 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 289.04 s
[root@jgao-test1 ~]# avocado run --vt-type libvirt type_specific.io-github-autotest-libvirt.virtual_disks.luks.negative_test.duplicated_encryption.coldplug.device_disk.dir_pool_test.target_format_qcow2
JOB ID     : 795bf478085fb0f74fb3cea03174c3e22b2ccbf7
JOB LOG    : /root/avocado/job-results/job-2020-04-02T22.47-795bf47/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virtual_disks.luks.negative_test.duplicated_encryption.coldplug.device_disk.dir_pool_test.target_format_qcow2.encryption_in_source: PASS (121.96 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 124.30 s
[root@jgao-test1 ~]# avocado run --vt-type libvirt type_specific.io-github-autotest-libvirt.virtual_disks.luks.negative_test.wrong_password.coldplug.device_disk.dir_pool_test.target_format_qcow2
JOB ID     : b4fcbcd526f2205e74329b9e4b0c75358fdee988
JOB LOG    : /root/avocado/job-results/job-2020-04-02T22.57-b4fcbcd/job.log
 (1/2) type_specific.io-github-autotest-libvirt.virtual_disks.luks.negative_test.wrong_password.coldplug.device_disk.dir_pool_test.target_format_qcow2.encryption_in_source: PASS (164.87 s)
 (2/2) type_specific.io-github-autotest-libvirt.virtual_disks.luks.negative_test.wrong_password.coldplug.device_disk.dir_pool_test.target_format_qcow2.encryption_out_source: PASS (184.63 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 352.12 s

```
Signed-off-by: jgao <jgao@localhost.localdomain>

### CODE REVIEW

File: libvirt/tests/src/virtual_disks/virtual_disks_luks.py [hunk 1]
```diff
@@ -374,9 +374,17 @@ def get_secret_list():
             vol_encryption_params.update({"format": "luks"})
             vol_encryption_params.update({"secret": {"type": "passphrase", "uuid": luks_sec_uuid}})
             try:
-                # If Libvirt version is lower than 2.5.0
-                # Creating luks encryption volume is not supported,so skip it.
-                create_vol(pool_name, vol_encryption_params, vol_params)
+                # If target format is qcow2,need to create test image with "qemu-img create"
+                if volume_target_format == "qcow2":
+                    password = params.get("luks_encrypt_passwd", "password")
+                    cmd = ("qemu-img create -f qcow2 "
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_luks.py:380]: In libvirt.create_local_disk(), it is supported to qemu-img create  -f ... an image. Could you help enhance this function with extra parameter like you use here, and then modify this to invoke create_local_disk() function. Your enhancement will also benefit others. Thanks.

File: libvirt/tests/cfg/virtual_disks/virtual_disks_luks.cfg [hunk 2]
```diff
@@ -10,6 +10,7 @@
     virt_disk_device_target = "vdb"
     virt_disk_device_format = "raw"
     virt_disk_device_bus = "virtio"
+
```
Comment (dzhengfy) [inline libvirt/tests/cfg/virtual_disks/virtual_disks_luks.cfg:13]: Please remove this unnecessary blank line.

**Comment by jgao12345** [comment]
This pr is relying to avocado-vt pr now: https://github.com/avocado-framework/avocado-vt/pull/2508

File: libvirt/tests/src/virtual_disks/virtual_disks_luks.py [hunk 3]
```diff
@@ -397,7 +403,7 @@ def get_secret_list():
         disk_xml = Disk(type_name=device_type)
         disk_xml.device = device
         disk_xml.target = {"dev": device_target, "bus": device_bus}
-        driver_dict = {"name": "qemu", "type": device_format}
+        driver_dict = {"name": "qemu", "type": volume_target_format}
```
Comment (lento-sun) [inline libvirt/tests/src/virtual_disks/virtual_disks_luks.py:406]: if 'device_format' param not used anywhere, it should be removed.

File: libvirt/tests/src/virtual_disks/virtual_disks_luks.py [hunk 4]
```diff
@@ -374,9 +373,16 @@ def get_secret_list():
...
             try:
-                # If Libvirt version is lower than 2.5.0
-                # Creating luks encryption volume is not supported,so skip it.
-                create_vol(pool_name, vol_encryption_params, vol_params)
+                # If target format is qcow2,need to create test image with "qemu-img create"
+                if volume_target_format == "qcow2":
+                    password = params.get("luks_encrypt_passwd", "password")
+                    extra = '--object secret,id=sec0,data=`printf %s | base64`,format=base64' \
+                            ' -o encrypt.format=luks,encrypt.key-secret=sec0' % password
+                    libvirt.create_local_disk("file", extra, volume_target_path, disk_format="qcow2", size="10M")
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_luks.py:381]: It is better to adjust the parameter order.

File: libvirt/tests/src/virtual_disks/virtual_disks_luks.py [hunk 5]
```diff
@@ -374,9 +373,16 @@ def get_secret_list():
             vol_encryption_params.update({"format": "luks"})
             vol_encryption_params.update({"secret": {"type": "passphrase", "uuid": luks_sec_uuid}})
             try:
-                # If Libvirt version is lower than 2.5.0
-                # Creating luks encryption volume is not supported,so skip it.
-                create_vol(pool_name, vol_encryption_params, vol_params)
+                # If target format is qcow2,need to create test image with "qemu-img create"
+                if volume_target_format == "qcow2":
+                    password = params.get("luks_encrypt_passwd", "password")
+                    extra = '--object secret,id=sec0,data=`printf %s | base64`,format=base64' \
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_luks.py:379]: Move the 'extra' in cfg.

**Comment by qiankehan** [comment]
@jgao12345 Update the 2nd commit subject.
I think the subject `Merge remote-tracking branch 'upstream/master'` is caused by `git pull` in your branch. Use `git pull origin master` to update your master branch.
A more easy way is to use to [hub](https://github.com/github/hub) tool: `hub sync`.

**Comment by sathnaga** [comment]
@jgao12345 rebase and squash commits..

**Comment by jgao12345** [comment]
> @jgao12345 rebase and squash commits..

Thanks for your reminding , i have updated the commit info.

File: libvirt/tests/src/virtual_disks/virtual_disks_luks.py [hunk 6]
```diff
@@ -376,9 +375,14 @@ def get_secret_list():
...
-                # If Libvirt version is lower than 2.5.0
-                # Creating luks encryption volume is not supported,so skip it.
-                create_vol(pool_name, vol_encryption_params, vol_params)
+                # If target format is qcow2,need to create test image with "qemu-img create"
+                if volume_target_format == "qcow2":
+                    option = params.get("luks_extra_elements")
+                    libvirt.create_local_disk("file", volume_target_path, extra=option, disk_format="qcow2", size="10M")
+                else:
+                    # If Libvirt version is lower than 2.5.0
+                    # Creating luks encryption volume is not supported,so skip it.
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_disks/virtual_disks_luks.py:384]: Do you need to use libvirt_version.version_compare to do the judgement？

File: libvirt/tests/src/virtual_disks/virtual_disks_luks.py [hunk 7]
```diff
@@ -159,7 +159,6 @@ def get_secret_list():
     # Disk specific attributes.
     device = params.get("virt_disk_device", "disk")
     device_target = params.get("virt_disk_device_target", "vdd")
-    device_format = params.get("virt_disk_device_format", "raw")
```
[the remark this answers is not in the document]
Comment (jgao12345) [reply libvirt/tests/src/virtual_disks/virtual_disks_luks.py:162]: Delete this parameter in cfg.

File: libvirt/tests/src/virtual_disks/virtual_disks_luks.py (hunk 6 above)
Comment (jgao12345) [reply libvirt/tests/src/virtual_disks/virtual_disks_luks.py:384]: No,i don't need it to do the judgement,i think the logic now is well.

**Comment by chunfuwen** [comment]
@jgao12345 , can you run all those cases related to luks since it change some logic for previous cases?

**Comment by jgao12345** [comment]
> @jgao12345 , can you run all those cases related to luks since it change some logic for previous cases?

Thanks for your advice which make me to avoid a big mistake.
I wrongly delete the parameter "device_format" before which work in my new case,but will get error in the old cases.So i add it back and commit again with result below.

# avocado run --vt-type libvirt type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test
JOB ID     : 41fdc6ad26be8e145bbcb08e67e557c168b1a909
JOB LOG    : /root/avocado/job-results/job-2020-05-19T10.53-41fdc6a/job.log
 (01/10) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test.enable_auth.auth_in_source.use_auth_uuid.encryption_in_source: PASS (75.24 s)
 (02/10) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test.enable_auth.auth_in_source.use_auth_uuid.encryption_out_source: PASS (83.95 s)
 (03/10) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test.enable_auth.auth_in_source.use_auth_usage.encryption_in_source: PASS (86.18 s)
 (04/10) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test.enable_auth.auth_in_source.use_auth_usage.encryption_out_source: PASS (85.46 s)
 (05/10) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test.enable_auth.auth_out_source.use_auth_uuid.encryption_in_source: PASS (85.24 s)
 (06/10) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test.enable_auth.auth_out_source.use_auth_uuid.encryption_out_source: PASS (84.89 s)
 (07/10) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test.enable_auth.auth_out_source.use_auth_usage.encryption_in_source: PASS (84.18 s)
 (08/10) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test.enable_auth.auth_out_source.use_auth_usage.encryption_out_source: PASS (85.01 s)
 (09/10) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test.disable_auth.encryption_in_source: PASS (84.66 s)
 (10/10) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.iscsi_test.disable_auth.encryption_out_source: PASS (84.56 s)
RESULTS    : PASS 10 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 842.50 s

# avocado run --vt-type libvirt type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.dir_pool_test
JOB ID     : 77896480d63b7e2b55921acac1dc3891c1dff3b0
JOB LOG    : /root/avocado/job-results/job-2020-05-19T22.06-7789648/job.log
 (1/4) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.dir_pool_test.target_format_raw.encryption_in_source: PASS (64.71 s)
 (2/4) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.dir_pool_test.target_format_raw.encryption_out_source: PASS (55.60 s)
 (3/4) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.dir_pool_test.target_format_qcow2.encryption_in_source: PASS (65.51 s)
 (4/4) type_specific.io-github-autotest-libvirt.virtual_disks.luks.positive_test.coldplug.device_disk.dir_pool_test.target_format_qcow2.encryption_out_source: PASS (65.90 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 254.19 s
# avocado run --vt-type libvirt type_specific.io-github-autotest-libvirt.virtual_disks.luks.negative_test.wrong_password.coldplug.device_disk.dir_pool_test
JOB ID     : 7399532371b56a48c477891e13a34a600a20d9af
JOB LOG    : /root/avocado/job-results/job-2020-05-19T22.19-7399532/job.log
 (1/4) type_specific.io-github-autotest-libvirt.virtual_disks.luks.negative_test.wrong_password.coldplug.device_disk.dir_pool_test.target_format_raw.encryption_in_source: PASS (54.19 s)
 (2/4) type_specific.io-github-autotest-libvirt.virtual_disks.luks.negative_test.wrong_password.coldplug.device_disk.dir_pool_test.target_format_raw.encryption_out_source: PASS (63.12 s)
 (3/4) type_specific.io-github-autotest-libvirt.virtual_disks.luks.negative_test.wrong_password.coldplug.device_disk.dir_pool_test.target_format_qcow2.encryption_in_source: PASS (39.59 s)
 (4/4) type_specific.io-github-autotest-libvirt.virtual_disks.luks.negative_test.wrong_password.coldplug.device_disk.dir_pool_test.target_format_qcow2.encryption_out_source: PASS (39.53 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 198.98 s

**Pull request merged**